sycl/include/sycl/accessor.hpp

```diff
@@ -2709,6 +2709,9 @@ class __SYCL_SPECIAL_CLASS local_accessor_base :
   bool operator!=(const local_accessor_base &Rhs) const {
     return !(*this == Rhs);
   }
+
+private:
+  template <typename DataT_, int Dimensions_> friend class local_accessor;
```

Comment:
Why does it need to be a friend? It inherits from it and isn't `impl` a protected member?

Reply:
> Why does it need to be a friend? It inherits from it and isn't `impl` a protected member?

the inheritance does not work as intended when the classes are specialised with differently.

Reply:
Could we potentially just use `detail::getSyclObjImpl` instead for those cases?

Reply:
great idea, I give it a try.

Reply:
attempt to replace `Rhs.impl` with `detail::getSyclObjImpl(Rhs)` in both `local_accessor` and `local_accessor_base` resulted in below assert.
```
common.hpp:301:10: error: value of type 'const detail::LocalAccessorBaseDevice<AdjustedDim>' is not contextually convertible to 'bool'
  assert(SyclObject.impl && "every constructor should create an impl"); 
```